iox_catalog/src/interface.rs

```diff
@@ -1214,26 +1298,199 @@ pub(crate) mod test_helpers {
         // remove namespace to avoid it from affecting later tests
         repos
             .namespaces()
-            .delete("test_namespace")
+            .soft_delete("test_namespace")
             .await
             .expect("delete namespace should succeed");
         repos
             .namespaces()
-            .delete("test_namespace2")
+            .soft_delete("test_namespace2")
             .await
             .expect("delete namespace should succeed");
         repos
             .namespaces()
-            .delete("test_namespace3")
+            .soft_delete("test_namespace3")
             .await
             .expect("delete namespace should succeed");
         repos
             .namespaces()
-            .delete("test_namespace4")
+            .soft_delete("test_namespace4")
             .await
             .expect("delete namespace should succeed");
     }
 
+    /// Construct a set of two namespaces:
+    ///
+    ///  * deleted-ns: marked as soft-deleted
+    ///  * active-ns: not marked as deleted
+    ///
+    /// And assert the expected "soft delete" semantics / correctly filter out
+    /// the expected rows for all three states of [`SoftDeletedRows`].
+    async fn test_namespace_soft_deletion(catalog: Arc<dyn Catalog>) {
+        let mut repos = catalog.repositories().await;
+        let topic = repos.topics().create_or_get("foo").await.unwrap();
+        let pool = repos.query_pools().create_or_get("foo").await.unwrap();
+
+        let deleted_ns = repos
+            .namespaces()
+            .create("deleted-ns", None, topic.id, pool.id)
+            .await
+            .unwrap();
+        let active_ns = repos
+            .namespaces()
+            .create("active-ns", None, topic.id, pool.id)
+            .await
+            .unwrap();
+
+        // Mark "deleted-ns" as soft-deleted.
+        repos.namespaces().soft_delete("deleted-ns").await.unwrap();
+
+        // Which should be idempotent (ignoring the timestamp change - when
+        // changing this to "soft delete" it was idempotent, so I am preserving
+        // that).
+        repos.namespaces().soft_delete("deleted-ns").await.unwrap();
```

Comment:
The biggest thing I see, which may be fine, is that soft-deleted namespaces can't be recreated with the exact same name because the namespace uniqueness check still applies whether or not the namespace is marked as deleted. That is, if I add this here:

```suggestion
        repos.namespaces().soft_delete("deleted-ns").await.unwrap();
        
        // Soft-deleted namespaces can't be recreated.
        let err = repos
            .namespaces()
            .create("deleted-ns", None, topic.id, pool.id)
            .await
            .unwrap_err();

        assert!(matches!(err, Error::NameExists { .. }));
```

it passes. 

Is that acceptable for now?

Reply:
I will raise this in our internal slack as I think it is a usability question more than a code question

Reply:
Update is that the ` it will be an error to create a new namespace with the same name`  sounds desirable: https://influxdata.slack.com/archives/CRK9M8L5Q/p1676059590176299?thread_ts=1676044231.571219&cid=CRK9M8L5Q

Reply:
Yes, this is by design, in cloud2 the namespaces are always unique, regardless of the friendly string name given by a user.

In the future we can look into making namespaces reusable in IOx (I imagine this will be desirable as a stand-alone product) but definitely out of scope for now.